### PR TITLE
feat: --copy-labels and GCP support

### PR DESCRIPTION
**Pull request by joemiller**
This PR introduces two new features:

1. Support for Google Cloud Platform (GCP). Set with `-cloud gcp`. AWS (`-cloud aws`) is the default if not specified.
2. New `-copy-labels` flag. When used, this flag will copy the specified labels (optionally all labels with `'*'`) from the PVC to the cloud disk volume.

GCP labels have a different set of constraints than k8s labels and AWS tags. The biggest difference is that `.` and `/` are not allowed, so a k8s label like `app.kubernetes.io/name` will be converted to `app-kubernetes-io_name` when applied to a GCP PD volume.

Also some small refactors:
- Converted provisioner magic strings like `ebs.csi.aws.com` to constants like `AWS_EBS_CSI`
- opportunistically converted some if/else blocks to switch statements

### CODE REVIEW

**Comment by mtougeron** [comment]
Thanks for the contribution! I'm swamped this afternoon but I'll take a look this weekend.

**Comment by joemiller** [comment]
@mtougeron no problem at all. We realize it might be kind of a lot in a single PR.

**Comment by mtougeron** [comment]
sorry, had some family issues come up over the weekend, nothing major, but it did keep me from being able to review this.

File: main.go [hunk 1]
```diff
@@ -338,3 +359,13 @@ func parseCsv(value string) map[string]string {
...
 }
+
+func parseCopyLabels(copyLabelsString string) []string {
+	if copyLabelsString == "*" {
+		return []string{"*"}
+	}
+	if copyLabelsString == "" {
+		return []string{}
+	}
+	return strings.Split(copyLabelsString, ",")
```
Comment (mtougeron) [inline main.go:370]: what if the user enters `foo,*,bar`?

Comment (mtougeron) [inline main.go:370]: Should probably also check against `foo,,bar` (empty values in the list)

File: main.go [hunk 2]
```diff
@@ -159,25 +199,9 @@ func main() {
...
-	}
-	awsSession = createAWSSession(region)
-	if awsSession == nil {
-		err = fmt.Errorf("nil AWS session: %v", awsSession)
-		if err != nil {
-			log.Println(err.Error())
-		}
-		os.Exit(1)
+	if copyLabelsString != "" {
+		copyLabels = strings.Split(copyLabelsString, ",")
```
Comment (mtougeron) [inline main.go:203]: Did you mean to call `parseCopyLabels` here?

File: gcp.go [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+	if err != nil {
+		log.Errorf("failed to delete labels from PD: %s", err)
+		promActionsTotal.With(prometheus.Labels{"status": "error", "storageclass": storageclass}).Inc()
+		return
+	}
+
+	waitForCompletion := func(_ context.Context) (bool, error) {
+		resp, err := c.GetGCEOp(project, location, op.Name)
+		if err != nil {
+			return false, fmt.Errorf("failed to delete labels from PD %s: %s", disk.Name, err)
```
Comment (mtougeron) [inline gcp.go:147]: Should probably increment the prometheus counter here.

```
promActionsTotal.With(prometheus.Labels{"status": "error", "storageclass": storageclass}).Inc()
```

File: gcp.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+		}
+		return resp.Status == "DONE", nil
+	}
+	if err := wait.PollUntilContextTimeout(context.TODO(),
+		time.Second,
+		time.Minute,
+		false,
+		waitForCompletion); err != nil {
+		log.Errorf("delete label operation failed: %s", err)
```
Comment (mtougeron) [inline gcp.go:156]:
```
promActionsTotal.With(prometheus.Labels{"status": "error", "storageclass": storageclass}).Inc()
```

**Review by mtougeron — COMMENTED** [review]
A few minor things. I haven't looked at the tests yet but I'll do that later tonight or in the morning.

File: main.go (hunk 1 above)
Comment (joemiller) [reply main.go:370]: Briefly thought about it, but did not have a solution other than "treat a single `*` entry as special" and document that is how to enable wildcard mode.

In the case of `foo,*,bar` should we enable wildcard mode? I could see someone making a typo while trying to do `foo*,bar` (btw `foo*` won't work) by adding a comma and then having their more specific label keys ignored suddenly all label keys are copied, which might be a surprising outcome.

We originally thought we would use wildcard mode but find ourselves using a more specific allow-list instead (though in my testing wildcard mode looked fine for general use). We could drop support for the magic `*`. And either add a new bool `--copy-all-labels` now, or not support labels at all and if a compelling request comes up then add the new bool flag.

File: main.go (hunk 2 above)
Comment (joemiller) [reply main.go:203]: yes 🤦

File: gcp.go (hunk 3 above)
Comment (joemiller) [reply gcp.go:147]: This closure may be called multiple times until success by the `PollUtilContext()` func a few lines lower.

If we only increment the metric after all retries fail (line 156) then we would have a single increment per disk attempt. I think that may be more desirable but I'm not positive.

I'd suggest only incrementing the counter after all retries have been exhausted. thoughts?

Comment (joemiller) [reply gcp.go]: Updating here and line 156 would result in a increasing the failure counter 2x for a single disk label update.

File: main.go (hunk 1 above)
Comment (mtougeron) [reply main.go]: I think the simplest solution is to just check for the string either being exclusively `*` or not in it at all.

File: gcp.go (hunk 3 above)
Comment (mtougeron) [reply gcp.go]: ah right, I ready it as 2 separate cases but you are correct

**Comment by joemiller** [comment]
@mtougeron I think I covered everything. LMK if there's more to address! tnx

**Review by mtougeron — APPROVED** [review]
/lgtm

**Comment by mtougeron** [comment]
I cut the new image release but I got pulled into things today. I'll try and get the chart updates done during my lunch break as well but it might not be until tomorrow morning.

**Pull request merged**